### ISSUE TITLE
Nightly regression around HRTBs, likely connected to wf changes

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#28934]
The following code:

```rust
struct Parser<'i: 't, 't>(&'i u8, &'t u8);

impl<'i, 't> Parser<'i, 't> {
    fn parse_nested_block<F, T>(&mut self, parse: F) -> Result<T, ()>
        where for<'tt> F: FnOnce(&mut Parser<'i, 'tt>) -> T { panic!() }
        
    fn expect_exhausted(&mut self) -> Result<(), ()> { Ok(()) }
}

fn main() {
    let x = 0u8;
    Parser(&x, &x).parse_nested_block(|input| input.expect_exhausted()).unwrap();
}
```

compiles on 1.3 stable and 1.4 beta, but not on nightly. This seems likely related to [wf changes](https://github.com/rust-lang/rfcs/pull/1214).

**Comment by aturon** [comment]
cc @SimonSapin @nikomatsakis

**Comment by aturon** [comment]
cc @arielb1

**Comment by SimonSapin** [comment]
The above is a reduction of https://github.com/servo/rust-cssparser/commit/29f757d2a52e3dac5bfe5ebe5e220f05bbb1e601

**Comment by SimonSapin** [comment]
I can reproduce in rustc 1.5.0-nightly (d0cae14f6 2015-10-08) but not in rustc 1.5.0-dev (168a23ebe 2015-10-01). (The latter is what Servo uses at the moment.)

**Comment by aturon** [comment]
A quick scan of the intervening commits suggests https://github.com/rust-lang/rust/commit/603a75c8eaa8ee168e4333e4fba5eb782ed7192b may be related.

**Comment by arielb1** [comment]
The eta-reduced version fails on all versions of rustc:
```Rust
struct Parser<'i: 't, 't>(&'i u8, &'t u8);

impl<'i, 't> Parser<'i, 't> {
    fn parse_nested_block<F, T>(&mut self, parse: F) -> Result<T, ()>
        where for<'tt> F: FnOnce(&mut Parser<'i, 'tt>) -> T { panic!() }

    fn expect_exhausted(&mut self) -> Result<(), ()> { Ok(()) }
}

fn main() {
    let x = 0u8;
    Parser(&x, &x).parse_nested_block(Parser::expect_exhausted).unwrap();
}
```

```
<anon>:12:20: 12:64 error: type mismatch: the type `fn(&mut Parser<'_, '_>) -> core::result::Result<(), ()> {Parser<'i, 't>::expect_exhausted}` implements the trait `for<'r> core::ops::FnOnce<(&'r mut Parser<'_, '_>,)>`, but the trait `for<'r, 'tt> core::ops::FnOnce<(&'r mut Parser<'_, 'tt>,)>` is required (expected concrete lifetime, found bound lifetime parameter 'tt) [E0281]
<anon>:12     Parser(&x, &x).parse_nested_block(Parser::expect_exhausted).unwrap();
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<anon>:12:20: 12:64 error: type mismatch resolving `for<'r, 'tt> <fn(&mut Parser<'_, '_>) -> core::result::Result<(), ()> {Parser<'i, 't>::expect_exhausted} as core::ops::FnOnce<(&'r mut Parser<'_, 'tt>,)>>::Output == _`:
 expected bound lifetime parameter 'tt,
    found concrete lifetime [E0271]
<anon>:12     Parser(&x, &x).parse_nested_block(Parser::expect_exhausted).unwrap();
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

<anon>:12:20: 12:64 help: see the detailed explanation for E0271
error: aborting due to 2 previous errors
```

The wf fixes just bought it in line.

**Comment by arielb1** [comment]
The UFCS version fails on 1.4+ - this looks like a closure inference issue.

**Comment by arielb1** [comment]
This was not fixed by #29006

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by nikomatsakis** [comment]
This seems to boil down to a bug in the givens list for region inference. It is fixed by #29188 (at least, it is with the latest commit).
